development/gcp/cloud-run/kms-rotate/kms-rotate.go

```diff
@@ -0,0 +1,194 @@
+		http.Error(w, "key value is not provided", http.StatusBadRequest)
+		return
+	}
+	if conf.BucketName == "" {
+		log.Print("bucket value is missing")
+		http.Error(w, "bucket value is not provided", http.StatusBadRequest)
+		return
+	}
+
+	keyPath := "projects/" + conf.Project + "/locations/" + conf.Location + "/keyRings/" + conf.Keyring + "/cryptoKeys/" + conf.Key
+	cryptoKeyRequest := &kmspb.GetCryptoKeyRequest{Name: keyPath}
+	cryptoKey, err := kmsService.GetCryptoKey(ctx, cryptoKeyRequest)
+	if err != nil {
+		log.Printf("get crypto key: %v", err)
+		http.Error(w, "Couldn't get crypto key", http.StatusBadRequest)
+		return
+	}
+	primaryVersion := cryptoKey.GetPrimary()
+
+	keyIteratorRequest := &kmspb.ListCryptoKeyVersionsRequest{Parent: keyPath}
+	keyIterator := kmsService.ListCryptoKeyVersions(ctx, keyIteratorRequest)
+	var keyVersions []*kmspb.CryptoKeyVersion
+	enabledVersions := 0
+	for nextVer, err := keyIterator.Next(); err != iterator.Done; nextVer, err = keyIterator.Next() {
+		if err != nil && err != iterator.Done {
+			log.Printf("key version iterator: %v", err)
+			http.Error(w, "Couldn't iterate over key versions", http.StatusBadRequest)
+			return
+		}
+		keyVersions = append(keyVersions, nextVer)
+		if nextVer.State == kmspb.CryptoKeyVersion_ENABLED {
+			enabledVersions++
+		}
+	}
+	if enabledVersions < 2 {
+		log.Printf("Less than two enabled key versions, quitting")
+		return
+	}
+
+	bucket := storageService.Bucket(conf.BucketName)
+
+	// for all files in bucket dir
+	query := &storage.Query{}
+	if conf.BucketPrefix != "" {
+		query.Prefix = conf.BucketPrefix
+	}
+
+	bucketIterator := bucket.Objects(ctx, query)
+
+	for {
+		attrs, err := bucketIterator.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// decrypt
+		// encrypt with latest
+		o := bucket.Object(attrs.Name)
+		reader, err := o.NewReader(ctx)
+		if err != nil {
+			log.Fatal(err)
+		}
+		encryptedData, err := io.ReadAll(reader)
+		if err != nil {
+			log.Fatal(err)
+		}
+		reader.Close()
+
+		decryptRequest := &kmspb.DecryptRequest{Name: keyPath, Ciphertext: encryptedData}
+		decryptResponse, err := kmsService.Decrypt(ctx, decryptRequest)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		encryptRequest := &kmspb.EncryptRequest{Name: keyPath, Plaintext: decryptResponse.Plaintext}
+		encryptResponse, err := kmsService.Encrypt(ctx, encryptRequest)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		writer := bucket.Object(attrs.Name).NewWriter(ctx)
+		_, err = writer.Write(encryptResponse.Ciphertext)
+		if err != nil {
+			log.Fatal(err)
+		}
+		writer.Close()
+	}
+
+	// destroy old keys
+	for _, keyVersion := range keyVersions {
+		if keyVersion.Name != primaryVersion.Name {
+			destructionRequest := &kmspb.DestroyCryptoKeyVersionRequest{Name: keyVersion.Name}
+			_, err := kmsService.DestroyCryptoKeyVersion(ctx, destructionRequest)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
```

Comment:
it could be a separate function and then you don't need a comment what it does.

Reply:
I would love to have a function documented too.